test/Vault1Test.js

```diff
@@ -2,12 +2,113 @@ const { expect } = require("chai");
+    const initialToken2Amount = await token2.balanceOf(owner.address);
+
+    expect(await vault1.tokenBalanceOf(owner.address, token1.address)).to.equal(
+      0
+    );
+
+    // deposits token1 into vault
+    await token1.connect(owner).approve(vault1.address, 1000);
+    await vault1.connect(owner).deposit(100, token1.address);
+    expect(await vault1.tokenBalanceOf(owner.address, token1.address)).to.equal(
+      100
+    );
+    expect(await token1.balanceOf(owner.address)).to.equal(
+      initialToken1Amount - 100
+    );
+
+    await token2.connect(owner).approve(vault1.address, 1000);
+    // deposits token2 into vault
+    await vault1.connect(owner).deposit(200, token2.address);
+    expect(await vault1.tokenBalanceOf(owner.address, token2.address)).to.equal(
+      200
+    );
+    expect(await token2.balanceOf(owner.address)).to.equal(
+      initialToken2Amount - 200
+    );
+  });
+
+  it("should withdraw ERC20 into vault", async () => {
+    // deposits token1 into vault
+    await token1.connect(owner).approve(vault1.address, 1000);
+    await vault1.connect(owner).deposit(100, token1.address);
+    const token1AmountOfOwner = await token1.balanceOf(owner.address);
+    await vault1.connect(owner).withdraw(50, token1.address);
+    expect(await vault1.tokenBalanceOf(owner.address, token1.address)).to.equal(
+      50
+    );
+    expect(await token1.balanceOf(owner.address)).to.equal(
+      Number(token1AmountOfOwner) + 50
+    );
+
+    // deposits token1 into vault
+    await token2.connect(owner).approve(vault1.address, 1000);
+    await vault1.connect(owner).deposit(1000, token2.address);
+    const token2AmountOfOwner = await token2.balanceOf(owner.address);
+    await vault1.connect(owner).withdraw(750, token2.address);
+    expect(await vault1.tokenBalanceOf(owner.address, token2.address)).to.equal(
+      250
+    );
+    expect(await token2.balanceOf(owner.address)).to.equal(
+      Number(token2AmountOfOwner) + 750
+    );
+  });
+
+  it("withdrawing the exact amount should be possible", async () => {
+    // deposits token1 into vault
+    await token1.connect(owner).approve(vault1.address, 1000);
+    await vault1.connect(owner).deposit(100, token1.address);
+    expect(await vault1.tokenBalanceOf(owner.address, token1.address)).to.equal(
+      100
+    );
+    await vault1.connect(owner).withdraw(100, token1.address);
+    expect(await vault1.tokenBalanceOf(owner.address, token1.address)).to.equal(
+      0
+    );
+    // should be reverted with underflow
+    await expect(vault1.connect(owner).withdraw(500, token1.address)).to.be
+      .reverted;
+  });
+
+  it("withdrawing more than it's available, should revert", async () => {
+    // deposits token1 into vault
+    await token1.connect(owner).approve(vault1.address, 1000);
+    await vault1.connect(owner).deposit(100, token1.address);
+    // should be reverted with underflow
+    await expect(vault1.connect(owner).withdraw(500, token1.address)).to.be
```

Comment:
This can be checked in contract using a revert condition and better to give out a meaningful revert string

Reply:
![image](https://user-images.githubusercontent.com/12175854/150949795-80ae4276-b0e4-4452-a328-784f196b0c10.png)
That's what I am trying to do here. I wanted to use in the test `revertedWith` the error message I am using, however it gave me an underflow error, hence I am using only `reverted`